design-documents/tracking-tool.md

```diff
@@ -0,0 +1,83 @@
+At least, we should have a tool which can find the latest version based on provided list and criteria.
+
+### Requirements
+
+ - An ability to provide the list of components for update
+ - An ability to recognized dependencies in `composer.json`
+ - Find the latest version according to the specific rule in multiple sources
+ - A possibility to extend the packages source list (Packagist, Github, Gitlab)
+ - Specific update rules like `^`, `~`, `-`, `*`, etc.
+ - A notification mechanism for available updates
+ 
+### Packages source 
+
+The most of packages are specified in `composer.json` but some of them are not specified, like database, or can be added in the future not via composer.
+So we need to provide a possibility to extend source of packages versions.
+
+The tool should support multiple input types like CLI, configuration files, URL to composer.json file.
+
+[Packagist](https://packagist.org) hosts all composer dependencies but, unfortunately, [Packagist API](https://packagist.org/apidoc) does not provide a possibility to retrieve the package latest version.
+But `https://packagist.org/packages/:vendor/:package-name.json` contains all released versions which could be parsed according to the specified rule.
+
+#### Github REST API
+
+As the most of packages (not only from `composer.json`) are hosted on Github, [Github REST API](https://developer.github.com/v3/repos/releases/) might be used to get components releases.
+
+Github provides different types of authentication:
+ - unauthenticated requests (60 requests per hour associated by IP address)
+ - authenticated requests (5000 requests per hour)
+ 
+Authenticated requests require multiple headers:
+
+ - `Authorization` should contain `token <token>` with permissions for public repositories
+ - `User-Agent` can be username or application ID
+ 
+Github provides multiple REST API entrypoints to get releases:
+
+ - `/repos/:ownwer/:repo/releases` - get a list of all published releases
+ - `/repos/:owner/:repo/latest` - get the latest published release
+ - `/repos/:owner/:repo/releases/:id` - get release details
+ - `/repos/:owner/:repo/tags` - get all repository tags
+ 
+The `/repos/:owner/:repo/latest` allows getting all needed details but the problem that not all, event popular, organizations publish their releases (as example, https://github.com/symfony/console/releases, https://github.com/php/php-src/releases) which are available on Github but not via REST API.
+
+The `/repos/:owner/:repo/tags` allows retrieving all tags (in the most cases, tags and releases the same) but the tags will be returned unsorted and some cases a repository contains not only versions tags.
+
+WEB crawlers can be used as a possible solution for repositories which do not publish releases as Github shows sorted by date releases based on tags list. The tool can try using Github REST API, if result is an empty array, it will use WEB crawler to parse the releases page.
```

Comment:
Tools that scrape the DOM of a site are always going to be incredibly flaky. I highly recommend avoiding scraping pages. The Github DOM has no versioning guarantees and can break at any time.

Reply:
You are right but as an alternative solution, it has the right to exist as there is no other API to get such details.

Reply:
there is the composer repo api, to get sorted version information.
You can also access that on the command line, e.g.:
composer show -a symfony/console

Or accessing the packagist api, by just adding .json to the packagist package url:
https://packagist.org/packages/symfony/console.json

I think the versions are already sorted, you only have to filter out dev versions